subsys/dfu/boot/mcuboot.c

```diff
@@ -0,0 +1,263 @@
+		*offs = IMAGE_OK_OFFS(bank_offs);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+
+static int boot_flash_write(off_t offs, const void *data, size_t len)
+{
+	int rc;
+
+	rc = flash_write_protection_set(flash_dev, false);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_write(flash_dev, offs, data, len);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_flag_write(int flag, u32_t bank_offs)
+{
+	u8_t buf[FLASH_MIN_WRITE_SIZE];
+	u32_t offs;
+	int rc;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	memset(buf, BOOT_FLAG_UNSET, sizeof(buf));
+	buf[0] = BOOT_FLAG_SET;
+
+	rc = boot_flash_write(offs, buf, sizeof(buf));
+
+	return rc;
+}
+
+static int boot_flag_read(int flag, u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+	u8_t flag_val;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = flash_read(flash_dev, offs, &flag_val, sizeof(flag_val));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return flag_val;
+}
+
+static int boot_image_ok_read(u32_t bank_offs)
+{
+	return boot_flag_read(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_image_ok_write(u32_t bank_offs)
+{
+	return boot_flag_write(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_magic_write(u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+
+	rc = boot_flash_write(offs, boot_img_magic, BOOT_MAGIC_SZ);
+
+	return rc;
+}
+
+static int boot_magic_code_check(const u32_t *magic)
+{
+	int i;
+
+	if (memcmp(magic, boot_img_magic, sizeof(boot_img_magic)) == 0) {
+		return BOOT_MAGIC_GOOD;
+	}
+
+	for (i = 0; i < 4; i++) {
+		if (magic[i] != 0xffffffff) {
+			return BOOT_MAGIC_BAD;
+		}
+	}
+
+	return BOOT_MAGIC_UNSET;
+}
+
+static int boot_magic_state_read(u32_t bank_offs)
+{
+	u32_t magic[4];
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+	rc = flash_read(flash_dev, offs, magic, sizeof(magic));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return boot_magic_code_check(magic);
+}
+
+int boot_request_upgrade(int permanent)
+{
+	int rc;
+
+	rc = boot_magic_write(FLASH_BANK1_OFFSET);
+	if (rc == 0 && permanent) {
+		rc = boot_image_ok_write(FLASH_BANK1_OFFSET);
+	}
+
+	return rc;
+}
+
+int boot_img_confirmed_write(void)
+{
+	int rc;
+
+	switch (boot_magic_state_read(FLASH_BANK0_OFFSET)) {
+	case BOOT_MAGIC_GOOD:
+		/* Confirm needed; proceed. */
+		break;
+
+	case BOOT_MAGIC_UNSET:
+		/* Already confirmed. */
+		return 0;
+
+	case BOOT_MAGIC_BAD:
+		/* Unexpected state. */
+		return -EFAULT;
+	}
+
+	if (boot_image_ok_read(FLASH_BANK0_OFFSET) != BOOT_FLAG_UNSET) {
+		/* Already confirmed. */
+		return 0;
+	}
+
+	rc = boot_image_ok_write(FLASH_BANK0_OFFSET);
+
+	return rc;
+}
+
+int boot_img_bank_erase(u32_t bank_offset)
+{
+	int rc;
+
+	rc = flash_write_protection_set(flash_dev, false);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_erase(flash_dev, bank_offset, FLASH_BANK_SIZE);
+	if (rc) {
+		return rc;
```

Comment:
This is tricky: you still want to try to write-protect the flash here instead of returning right away.

Reply:
The API says: "on some flash components, the write protection is automatically turned on again by the device after the completion of each write or erase calls"

The operative word here is "completion", not "successful completion."

Reply:
Do you mean that flash should be protected back even in case erase operation had failed?